File: plugin/integration/scala-plugin/src/scala/org/intellij/plugin/zeppelin/scala/runner/ZeppelinCustomRunner.scala
package org.intellij.plugin.zeppelin.scala.runner

import org.intellij.plugin.zeppelin.components.ZeppelinComponent
import org.jetbrains.plugins.scala.worksheet.processor.{WorksheetCompilerUtil, WorksheetCustomRunner}

/**
  * A Zeppelin cell runner
  */
class ZeppelinCustomRunner extends WorksheetCustomRunner {
  override def canHandle(request: WorksheetCompilerUtil.RunCustom): Boolean = {
    val id = request.id.split("\n")(0)
    id == ZeppelinCustomRunner.RUNNER_ID
  }

  override def handle(request: WorksheetCompilerUtil.RunCustom): Unit = {
    val executeCode = request.data
    val id = parseId(request.id)
    val connection = ZeppelinComponent.connectionFor(request.project)
    connection.focusToLog()
    val service = connection.service
    service.runCode(executeCode, id)
  }

  def parseId(s: String): Option[String] = {
    val ids = s.split("\n")
    if (ids.length<2) return  None
    val rawId = ids(1)
    if (rawId.isEmpty) None
    Some(rawId)
  }
}

object ZeppelinCustomRunner {
  val RUNNER_ID = "Zeppelin"
}

File: zeppelin/integration/scala-plugin/src/org/intellij/plugin/zeppelin/scala/runner/ZeppelinCustomRunner.kt
package org.intellij.plugin.zeppelin.scala.runner

import org.intellij.plugin.zeppelin.components.ZeppelinComponent
import org.jetbrains.plugins.scala.worksheet.processor.WorksheetCompilerUtil.RunCustom
import org.jetbrains.plugins.scala.worksheet.processor.WorksheetCustomRunner

/**
 * A Zeppelin cell runner
 */
open class ZeppelinCustomRunner : WorksheetCustomRunner() {
    override fun canHandle(request: RunCustom): Boolean {
        val id: String = request.id().split("\n")[0]
        return id == ZeppelinCustomRunner.RUNNER_ID
    }

    override fun handle(request: RunCustom) {
        val executeCode: String = request.data()
        val id: String? = parseId(request.id())
        val connection: ZeppelinComponent = ZeppelinComponent.connectionFor(request.project())
        connection.focusToLog()
        val service = connection.service
        service.runCode(executeCode, id)
    }

    private fun parseId(s: String): String? {
        val ids: List<String> = s.split("\n")
        if (ids.size < 2) return null
        val rawId: String = ids[1]
        if (rawId.isEmpty()) return null
        return rawId
    }

    companion object {
        const val RUNNER_ID: String = "Zeppelin"
    }
}